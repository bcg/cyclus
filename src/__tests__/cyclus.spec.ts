import { Lifecycle, SystemMap, using } from "../cyclus";

describe("cyclus", () => {
  class Database extends Lifecycle {
    dbConnection: string;

    start() {
      order.push("Start Database");
      this.dbConnection = "OPENED";
    }

    stop() {
      order.push("Stop Database");
      this.dbConnection = "CLOSED";
    }
  }

  class Scheduler extends Lifecycle {
    tick: number;

    start() {
      order.push("Start Scheduler");
      this.tick = 10;
    }

    stop() {
      order.push("Stop Scheduler");
    }
  }
  const createDatabase = (): Lifecycle => new Database();
  const createScheduler = (): Lifecycle => new Scheduler();
  let order: Array<string>;
  let system: SystemMap;

  beforeEach(() => {
    order = [];
  });

  describe("lifecycle", () => {
    class ExampleComponent extends Lifecycle {
      database: Database;
      scheduler: Scheduler;

      start() {
        order.push("Start ExampleComponent");
      }

      stop() {
        order.push("Stop ExampleComponent");
      }
    }
    const createExampleComponent = (): Lifecycle => new ExampleComponent();

    beforeEach(() => {
      system = new SystemMap({
        database: createDatabase(),
        scheduler: createScheduler(),
        exampleComponent: using(createExampleComponent(), [
          "database",
          "scheduler"
        ])
      });
      system.start();
    });

    it("should start the system in order", () => {
      expect(order).toMatchSnapshot();
    });

    it("should stop the system in reversed order", () => {
      system.stop();
      expect(order).toMatchSnapshot();
    });

    it("starting should be idempotent", () => {
      system.start();
      expect(order).toMatchSnapshot();
    });

    it("stopping should be idempotent", () => {
      system.stop();
      system.stop();
      expect(order).toMatchSnapshot();
    });
  });

  describe("injecting dependencies correctly", () => {
    class ExampleComponent extends Lifecycle {
      db: Database;
      sched: Scheduler;

      start() {
        order.push("Start ExampleComponent");
      }

      stop() {
        order.push("Stop ExampleComponent");
      }
    }
    const createExampleComponent = (): Lifecycle => new ExampleComponent();

    it("'should work with map", () => {
      system = new SystemMap({
        database: createDatabase(),
        scheduler: createScheduler(),
        exampleComponent: using(createExampleComponent(), {
          database: "db",
          scheduler: "sched"
        })
      });

      system.start();

      expect(system).toMatchSnapshot();
    });
  });

  describe("replacing dependencies on the fly", () => {
    class NewDatabase extends Lifecycle {
      dbConnection: string;

      start() {
        order.push("Start NewDatabase");
        this.dbConnection = "OPENED";
      }

      stop() {
        order.push("Stop NewDatabase");
        this.dbConnection = "CLOSED";
      }
    }

    class ExampleComponent extends Lifecycle {
      database: Database | NewDatabase;
      scheduler: Scheduler;

      start() {
        order.push("Start ExampleComponent");
      }

      stop() {
        order.push("Stop ExampleComponent");
      }
    }
    const createNewDatabase = (): Lifecycle => new NewDatabase();
    const createExampleComponent = (): Lifecycle => new ExampleComponent();

    it("'should work correctly", () => {
      system = new SystemMap({
        database: createDatabase(),
        scheduler: createScheduler(),
        exampleComponent: using(createExampleComponent(), [
          "database",
          "scheduler"
        ])
      });

      system.start();

      expect(system).toMatchSnapshot();
      system.replace({ database: createNewDatabase() });
      expect(system).toMatchSnapshot();

      expect(order).toMatchSnapshot();
    });
  });
});
