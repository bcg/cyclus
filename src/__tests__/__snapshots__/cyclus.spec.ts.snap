// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cyclus injecting dependencies correctly 'should work with map 1`] = `
SystemMap {
  "map": Object {
    "database": Database {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
      "dbConnection": "OPENED",
    },
    "exampleComponent": ExampleComponent {
      "__metadata": Object {
        "dependencies": Object {
          "database": "db",
          "scheduler": "sched",
        },
        "isInitialised": true,
      },
      "db": Database {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
        "dbConnection": "OPENED",
      },
      "sched": Scheduler {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
        "tick": 10,
      },
    },
    "scheduler": Scheduler {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
      "tick": 10,
    },
  },
  "order": Array [
    "scheduler",
    "database",
    "exampleComponent",
  ],
}
`;

exports[`cyclus lifecycle should start the system in order 1`] = `
Array [
  "Start Scheduler",
  "Start Database",
  "Start ExampleComponent",
]
`;

exports[`cyclus lifecycle should stop the system in reversed order 1`] = `
Array [
  "Start Scheduler",
  "Start Database",
  "Start ExampleComponent",
  "Stop ExampleComponent",
  "Stop Database",
  "Stop Scheduler",
]
`;

exports[`cyclus lifecycle starting should be idempotent 1`] = `
Array [
  "Start Scheduler",
  "Start Database",
  "Start ExampleComponent",
]
`;

exports[`cyclus lifecycle stopping should be idempotent 1`] = `
Array [
  "Start Scheduler",
  "Start Database",
  "Start ExampleComponent",
  "Stop ExampleComponent",
  "Stop Database",
  "Stop Scheduler",
]
`;

exports[`cyclus replacing dependencies on the fly 'should work correctly 1`] = `
SystemMap {
  "map": Object {
    "database": Database {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
      "dbConnection": "OPENED",
    },
    "exampleComponent": ExampleComponent {
      "__metadata": Object {
        "dependencies": Object {
          "database": "database",
          "scheduler": "scheduler",
        },
        "isInitialised": true,
      },
      "database": Database {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
        "dbConnection": "OPENED",
      },
      "scheduler": Scheduler {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
        "tick": 10,
      },
    },
    "scheduler": Scheduler {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
      "tick": 10,
    },
  },
  "order": Array [
    "scheduler",
    "database",
    "exampleComponent",
  ],
}
`;

exports[`cyclus replacing dependencies on the fly 'should work correctly 2`] = `
SystemMap {
  "map": Object {
    "database": NewDatabase {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
      "dbConnection": "OPENED",
    },
    "exampleComponent": ExampleComponent {
      "__metadata": Object {
        "dependencies": Object {
          "database": "database",
          "scheduler": "scheduler",
        },
        "isInitialised": true,
      },
      "database": NewDatabase {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
        "dbConnection": "OPENED",
      },
      "scheduler": Scheduler {
        "__metadata": Object {
          "dependencies": Object {},
          "isInitialised": true,
        },
        "tick": 10,
      },
    },
    "scheduler": Scheduler {
      "__metadata": Object {
        "dependencies": Object {},
        "isInitialised": true,
      },
      "tick": 10,
    },
  },
  "order": Array [
    "scheduler",
    "database",
    "exampleComponent",
  ],
}
`;

exports[`cyclus replacing dependencies on the fly 'should work correctly 3`] = `
Array [
  "Start Scheduler",
  "Start Database",
  "Start ExampleComponent",
  "Stop Database",
  "Start NewDatabase",
]
`;
